cypress-io/cypress
# chore: Implement runSpec mutation

- Closes cypress-io/cypress-internal#593

### Additional details

New GQL endpoint to accept a specPath and, given the matching project is already open, switch to the necessary testingType and launch matching spec. Eventual goal of this endpoint is to accept a project path and open project as necessary as well as a browser arg but those are out of scope for this ticket.

Mutation will validate:
* A project is open
* specPath is non-empty
* specPath points to an existing file
* specPath matches either e2e or ct specPattern
* matching testing type is configured
* Specified path points to a valid spec within the testing type
* A browser is available to launch into

Endpoint will return either a success response with info on the testingType, spec, and browser
...or an error response with a `code` field indicating the specific error type

### Steps to test
1. Find a project with e2e and CT configured, pick a spec from one mode or the other and record the relative path from project root
2. Open in global mode
3. Launch GraphiQL `http://localhost:4444/__launchpad/graphql`
4. Execute `runSpec` mutation with spec path from above - should fail since no project is open and give appropriate error code
```
mutation {
  runSpec (specPath: "cypress/e2e/1-getting-started/todo.cy.js") {
    __typename
    ... on RunSpecResponse {
      testingType
      browser {
        id
        name
      }
      spec {
        id
        name
      }
    }
    ... on RunSpecError {
      code
      detailMessage
    }
  }
}
```
5. Open selected project into testing mode matching the selected spec
6. Execute `runSpec` - should successfully launch spec and return a 'SUCCESS' response
7. Provide an empty specPath to mutation - should fail and give 'NO_SPEC_PATH' error response
8. Change specPath in mutation to point to a file that does not exist (but still satisfies specPattern) - should fail and give'SPEC_NOT_FOUND' error code
9. Change specPath to not match specPattern for either testingType - should fail and give 'NO_SPEC_PATTERN_MATCH' error code
10. Pick a spec from other testing type and use relative path in mutation - should successfully switch testing type and run spec, return valid SUCCESS response
11. Use devtools to record one of the gql requests, copy as a cURL command. Run cURL command from terminal to verify it is externally accessible
12. Close Cypress, edit project configuration to remove config for one testing type
13. Relaunch & open project. Use GQL to attempt to run a spec from now-unconfigured testing type. Should fail with a 'TESTING_TYPE_NOT_CONFIGURED' response

**Extra Credit**
Validate runSpec executes as expected when:
* Browser is open in matching testing mode
* Browser is open in other testing mode
* Cypress is on the launchpad (no browser open)

### How has the user experience changed?
No user-facing change

https://github.com/cypress-io/cypress/assets/11482842/25bdedc2-5358-4c15-945f-000049b48eec

### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
